### PR TITLE
feat(catalog-import): Make catalog filename and branch customizable

### PR DESCRIPTION
**Pull request by kmikko**
## Hey, I just made a Pull Request!

[Documentation](https://backstage.io/docs/features/software-catalog/descriptor-format) suggest using `catalog-info.yaml` as entity descriptor filename but as mentioned in ADR [here](https://github.com/backstage/backstage/blob/master/docs/architecture-decisions/adr008-default-catalog-file-name.md) it's not a requirement. Based on Discord discussion related to [this](https://discord.com/channels/687207715902193673/928654066538647552) matter and how core plugins [should](https://discord.com/channels/687207715902193673/687207715902193679/791036454767165480) be configurable enough I decided to give this a shot and implement it as a system configuration option.

The use case for this is as presented in Discord dicussion; our organization is already using custom filename for catalog entity descriptors which work fine with the automatic catalog ingestion loop. However catalog import (`plugin-catalog-import`) / register an existing component -flow uses a hardcoded value of `catalog-info.yaml` which we would like to also customize. 

This PR introduces configuration options for both catalog filename and the branch name used in pull requests. Besides these, optional props for `StepInitAnalyzeUrl` and `ImportInfoCard` are introduced to allow customization of the example paths. 
This is the part I'm a bit unsure of; as `StepInitAnalyzeUrl` is used deep in the state machine of `ImportStepper`. I wouldn't like to have to copy-paste the entire implementation just to customize this one example path. However, such as implemented providing this prop by using custom stepper with `ImportStepper` can be achieved with minimal effort. 
If you have alternative ideas I'm happy to hear them.

```tsx
const customStepper = (
  flow: ImportFlows,
  defaults: StepperProvider,
): StepperProvider => {
  const stepper = defaultGenerateStepper(flow, defaults);

  return {
    ...stepper,
    analyze: (state, { apis }) => ({
      stepLabel: <StepLabel>Select URL</StepLabel>,
      content: (
        <StepInitAnalyzeUrl
          key="analyze"
          analysisUrl={state.analysisUrl}
          onAnalysis={state.onAnalysis}
          disablePullRequest={!apis.catalogImportApi.preparePullRequest}
          exampleEntityFilepath="https://github.com/acme-corp/our-awesome-api/blob/main/anvil.yaml"
        />
      ),
    }),
  };
};

// App.tsx
const routes = (
    ...
    <Route path="/catalog-import" element={<CatalogImportPage />}>
      <Page themeId="home">
        <Header title="Register an existing component" />
        <Content>
          <ContentHeader title="Start tracking your components">
            <SupportButton>
              Start tracking your component in ACME Developer Portal by adding it to the
              software catalog.
            </SupportButton>
          </ContentHeader>

          <Grid container spacing={2} direction="row-reverse">
            <Grid item xs={12} md={4} lg={6} xl={8}>
              <ImportInfoCard
                exampleEntityFilepath="https://github.com/acme-corp/our-awesome-api/blob/main/anvil.yaml"
                exampleRepositoryPath="https://github.com/acme-corp/our-awesome-api"
              />
            </Grid>

            <Grid item xs={12} md={8} lg={6} xl={4}>
              <ImportStepper generateStepper={customStepper} />
            </Grid>
          </Grid>
        </Content>
      </Page>
    </Route>
);
```

Screenshots of the catalog import flow showcasing the customizations:

Screenshot 2022-01-14 at 13 15 44
Screenshot 2022-01-14 at 13 14 36
Screenshot 2022-01-14 at 13 15 24

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by kmikko** [comment]
Generated [updated](https://github.com/backstage/backstage/runs/4842617192?check_suite_focus=true) api-report and rebased against master

File: .changeset/tame-ads-exercise.md
Comment (freben) [inline .changeset/tame-ads-exercise.md:7]: these two can be removed

File: plugins/catalog-import/src/components/StepPrepareCreatePullRequest/StepPrepareCreatePullRequest.test.tsx
Comment (freben) [inline plugins/catalog-import/src/components/StepPrepareCreatePullRequest/StepPrepareCreatePullRequest.test.tsx:58]: don't mock this one, use `new TestConfigApi({ /* actual config structure as a json object */ })` from `@backstage/test-utils`

File: plugins/catalog-import/config.d.ts
Comment (freben) [inline plugins/catalog-import/config.d.ts:26]: Hm. Thinking about the longer term and scope creep, it may be wise to make these have longer and more specific names. Perhaps `defaultEntityFilename` here? This is an OK config parameter to have, I think.

File: plugins/catalog-import/config.d.ts
Comment (freben) [inline plugins/catalog-import/config.d.ts:33]: Hm a little bit less sure about this one. First instinct is that this could just be a prop on the frontend components instead because it's very unlikely that anybody except one single place would want to consume it. Either way, it may make sense to name it `defaultGeneratedEntityBranchName` or similar, if it should stay at all.

File: plugins/catalog-import/api-report.md
Comment (freben) [inline plugins/catalog-import/api-report.md:246]: I am pretty sure that we shouldn't export this component. It's highly internal and not stable enough to make a lot of sense for others to consume. And as you see, it pretty heavily pollutes the API surface - at the bottom you see warnings that even _more_ types and stuff would need to be exported as well to make this complete.

I say revert this one.

File: plugins/catalog-import/src/components/ImportInfoCard/ImportInfoCard.tsx
Comment (freben) [inline plugins/catalog-import/src/components/ImportInfoCard/ImportInfoCard.tsx:30]: Yeah these props are probably fine - they stay local to the place where they are actually needed. So what remains then is to make them easy for the user to set. What if they could be set on `DefaultImportPage`? That's already exposed publicly, and it can be responsible for passing them down. That means that the user would do like

```tsx
<Route path="/catalog-import" element={
  <CatalogImportPage>
    <DefaultImportPage exampleEntityFilepath="..." />
  </CatalogImportPage>
} />
```

Comment (freben) [reply plugins/catalog-import/src/components/ImportInfoCard/ImportInfoCard.tsx:30]: This would then also be where the user could pass in a branch name - making the config parameter for that unnecessary. I think it might not be particularly common to want to tweak, and its scope is really limited to within this component.

File: plugins/catalog-import/src/components/StepPrepareCreatePullRequest/StepPrepareCreatePullRequest.test.tsx
Comment (kmikko) [reply plugins/catalog-import/src/components/StepPrepareCreatePullRequest/StepPrepareCreatePullRequest.test.tsx:58]: I couldn't find any method named like that, did you mean `ConfigReader` from `@backstage/core-app-api`?
So something like:
```tsx
const configApi = new ConfigReader({
    catalog: {
      filename: undefined,
    },
  });
```

File: plugins/catalog-import/config.d.ts
Comment (kmikko) [reply plugins/catalog-import/config.d.ts:26]: Do you see this parameter being useful outside of catalog import? If not, should this be moved under `catalog.import.filename` for example?
Did a lot of back-and-forth with the name, how do you feel about `entityDescriptorFilename` or just `entityFilename`? I think having `default` in the name is obsolete as there's a default value for this parameter and if you end up changing it it's not default anymore.

File: plugins/catalog-import/src/components/StepPrepareCreatePullRequest/StepPrepareCreatePullRequest.test.tsx
Comment (freben) [reply plugins/catalog-import/src/components/StepPrepareCreatePullRequest/StepPrepareCreatePullRequest.test.tsx:58]: Yep that works fine too, they are basically equivalent but the other one is geared toward tests. You may be on an earlier version then of test-utils that doesn't have that yet

File: plugins/catalog-import/api-report.md
Comment (kmikko) [reply plugins/catalog-import/api-report.md:246]: I ended up exporting this because I needed the type definitions for custom stepper for `ImportStepper` component. I understand that the definitions might not be stable enough but that makes me wonder if `generateStepper` prop should be there after all? 

I guess I could do something like the following to provide `exampleEntityFilepath` and to rely on implicit types which wouldn't require exposing types used in `generateStepper`:
```tsx
<ImportStepper
  generateStepper={(flow, defaults) => {
    const stepper = defaultGenerateStepper(flow, defaults);

    return {
      ...stepper,
      analyze: (state, { apis }) => ({
        stepLabel: <StepLabel>Select URL</StepLabel>,
        content: (
          <StepInitAnalyzeUrl
            key="analyze"
            analysisUrl={state.analysisUrl}
            onAnalysis={state.onAnalysis}
            disablePullRequest={
              !apis.catalogImportApi.preparePullRequest
            }
            exampleEntityFilepath="https://github.com/acme-corp/our-awesome-api/blob/main/anvil.yaml"
          />
        ),
      }),
    };
  }}
/>
```
I understand this is very fragile but haven't come up with a better way to provide this example entity filepath without needing to re-write the whole stepper logic.

File: plugins/catalog-import/src/components/ImportInfoCard/ImportInfoCard.tsx
Comment (kmikko) [reply plugins/catalog-import/src/components/ImportInfoCard/ImportInfoCard.tsx:30]: I would love to have something like this! However as I see it leaves two problems:
- How to pass `exampleEntityFilepath` to `StepInitAnalyzeUrl`. The best I've come up with is to provide it via `generateStepper` to `ImportStepper`. Please see my other comment about this.
- Branch name is used in `CatalogImportClient` and I wouldn't want to create my own implementation just to patch this branch name in couple of places. Thus I thought having it as a config option would be the best option.

File: plugins/catalog-import/src/components/StepPrepareCreatePullRequest/StepPrepareCreatePullRequest.test.tsx
Comment (kmikko) [reply plugins/catalog-import/src/components/StepPrepareCreatePullRequest/StepPrepareCreatePullRequest.test.tsx:58]: Ah, re-checked the changelog for `@backstage/test-utils` and you probably meant `MockConfigApi` instead of `Test...`. I was previously only checking methods starting with `test` so missed it. Will use that instead.

File: plugins/catalog-import/config.d.ts
Comment (kmikko) [reply plugins/catalog-import/config.d.ts]: Settled on `pullRequestBranchName` for now

File: plugins/catalog-import/config.d.ts
Comment (kmikko) [reply plugins/catalog-import/config.d.ts]: Settled for `entityFilename`

File: plugins/catalog-import/api-report.md
Comment (kmikko) [reply plugins/catalog-import/api-report.md]: Reverted the change, `exampleEntityFilepath` can still be passed down to `StepInitAnalyzeUrl` as in the example above

File: plugins/catalog-import/src/components/StepPrepareCreatePullRequest/StepPrepareCreatePullRequest.tsx
Comment (Rugvip) [inline plugins/catalog-import/src/components/StepPrepareCreatePullRequest/StepPrepareCreatePullRequest.tsx:110]: This is repeated enough times that it's kinda worth making a local `useCatalogFilename()` hook 😅, or a `getCatalogFilename(config)` helper

File: plugins/catalog-import/src/components/StepPrepareCreatePullRequest/PreviewCatalogInfoComponent.test.tsx
Comment (Rugvip) [inline plugins/catalog-import/src/components/StepPrepareCreatePullRequest/PreviewCatalogInfoComponent.test.tsx:124]: Using the `<TestApiProvider>` here lets you use fewer characters :grin:

**Review by Rugvip — COMMENTED** [review]
Thank you, looks good! 👍 😁 

A thought on the config schema and some small ideas

File: plugins/catalog-import/src/components/StepPrepareCreatePullRequest/PreviewCatalogInfoComponent.test.tsx
Comment (kmikko) [reply plugins/catalog-import/src/components/StepPrepareCreatePullRequest/PreviewCatalogInfoComponent.test.tsx]: Nice! Converted all tests in this file to use it

File: plugins/catalog-import/src/components/StepPrepareCreatePullRequest/StepPrepareCreatePullRequest.tsx
Comment (kmikko) [reply plugins/catalog-import/src/components/StepPrepareCreatePullRequest/StepPrepareCreatePullRequest.tsx]: Good point, introduced both of them (and additional `getBranchName`) as these need to be accessed both in React components and in `CatalogImportApi`.

File: plugins/catalog-import/config.d.ts
Comment (kmikko) [reply plugins/catalog-import/config.d.ts:21]: Yeah I was also wondering the same thing if this would be helpful outside of the catalog import scope. Moved these under `catalog.import` to be more explicit.

**Comment by kmikko** [comment]
Thanks for the feedback! Made some adjustments and updated catalog import `README` about changes introduced in this PR. Let me know if there's more to it.

**Review by Rugvip — APPROVED** [review]
Thank you! 👍 

Looks good to me :shipit:
Just wanna double check with the rest of @backstage/maintainers before merge tomorrow as there are a couple of significant additions.

File: plugins/catalog-import/src/components/ImportInfoCard/ImportInfoCard.tsx
Comment (Rugvip) [inline plugins/catalog-import/src/components/ImportInfoCard/ImportInfoCard.tsx:26]: Alright, discussed, and we discovered one thing that we think is worth tweaking. What do you think about calling these `exampleLocationUrl` and `exampleRepositoryUrl` instead?

Comment (kmikko) [reply plugins/catalog-import/src/components/ImportInfoCard/ImportInfoCard.tsx:26]: I'm fine with the suggested name changes 👍

Comment (kmikko) [reply plugins/catalog-import/src/components/ImportInfoCard/ImportInfoCard.tsx]: and done

**Comment by kmikko** [comment]
Fixed conflict in the `package.json`

**Review by jhaals — APPROVED** [review]

**Comment by jhaals** [comment]
Thanks!

**Pull request merged**
